YNU-354: Protocol Docs (#69)
layer-3/docs: docs/protocol/off-chain/transfers.mdx

---
sidebar_position: 5

> title: Transfer Methods

Reviewer: there is only one transfer method

Author: [1eeca228bfcc771893fae0befb9d0101c0bd888f](https://github.com/layer-3/docs/pull/69/commits/1eeca228bfcc771893fae0befb9d0101c0bd888f)

---

import Tooltip from '@site/src/components/Tooltip';

# Transfer Methods

> Transfer methods enable instant, off-chain fund movement between users connected to the same <Tooltip content="A virtual ledger layer that provides a unified ledger (through Nitro RPC) and coordinates state channels (through Nitrolite), providing chain abstraction for developers and users">clearnode</Tooltip>.

Reviewer: I would not mention "connected to the same clearnode".

Author: [1eeca228bfcc771893fae0befb9d0101c0bd888f](https://github.com/layer-3/docs/pull/69/commits/1eeca228bfcc771893fae0befb9d0101c0bd888f)

---

## Overview

The transfer system allows users to send funds to each other instantly using their <Tooltip content="An abstraction that aggregates a user's funds across multiple blockchain networks, managed by Clearnode">unified balance</Tooltip>, without any on-chain transactions. Transfers are backed by the security of underlying payment channels and use double-entry bookkeeping for accounting accuracy.

### Why Use Transfers?

> **Instant Settlement**: Transfers complete in under 1 second with immediate finality.

Reviewer: I wouldn't specify "under 1 second". You may say Transfers complete immediately with instant finality.

Author: done

Author: [2407a50f77b866e3cddc84621a441c69e6787515>](https://github.com/layer-3/docs/pull/69/commits/2407a50f77b866e3cddc84621a441c69e6787515)

> **Zero Gas Fees**: No blockchain transactions means zero gas costs for both sender and recipient.

Reviewer: Make it "No blockchain fees" as some fees would be introdused at some point on the YN side

Author: [1eeca228bfcc771893fae0befb9d0101c0bd888f](https://github.com/layer-3/docs/pull/69/commits/1eeca228bfcc771893fae0befb9d0101c0bd888f)

> **Cross-Chain Unified**: Send from your aggregated balance across multiple chains without specifying which chain.

Reviewer: we can remove "without specifying which chain."

Author: [1eeca228bfcc771893fae0befb9d0101c0bd888f](https://github.com/layer-3/docs/pull/69/commits/1eeca228bfcc771893fae0befb9d0101c0bd888f)

**Auditable**: Complete transaction history with double-entry ledger tracking.

:::success Instant Off-Chain Payments
Transfers provide the speed and convenience of traditional payment networks while maintaining the security guarantees of blockchain-backed channels.
:::

---

## transfer

### Name

`transfer`

### Usage

> Transfers funds from the authenticated user's <Tooltip content="An abstraction that aggregates a user's funds across multiple blockchain networks, managed by Clearnode">unified balance</Tooltip> to another user's <Tooltip content="An abstraction that aggregates a user's funds across multiple blockchain networks, managed by Clearnode">unified balance</Tooltip> within the Yellow Network. This is a purely off-chain operation executed as a database transaction on <Tooltip content="A virtual ledger layer that provides a unified ledger (through Nitro RPC) and coordinates state channels (through Nitrolite), providing chain abstraction for developers and users">the clearnode</Tooltip>, resulting in instant settlement (typically under 1 second) with zero gas fees. The transfer updates internal ledger entries using double-entry bookkeeping principles and creates a transaction record for both parties. Both sender and recipient must be connected to the same <Tooltip content="A virtual ledger layer that provides a unified ledger (through Nitro RPC) and coordinates state channels (through Nitrolite), providing chain abstraction for developers and users">clearnode</Tooltip> (have open channels). The security guarantee comes from the underlying on-chain channels that back the <Tooltip content="An abstraction that aggregates a user's funds across multiple blockchain networks, managed by Clearnode">unified balance</Tooltip>.

Reviewer: "(have open channels)" - this is not right. having an open channel is not a prerequisite

Author: [1eeca228bfcc771893fae0befb9d0101c0bd888f](https://github.com/layer-3/docs/pull/69/commits/1eeca228bfcc771893fae0befb9d0101c0bd888f)

> Transfers funds from the authenticated user's <Tooltip content="An abstraction that aggregates a user's funds across multiple blockchain networks, managed by Clearnode">unified balance</Tooltip> to another user's <Tooltip content="An abstraction that aggregates a user's funds across multiple blockchain networks, managed by Clearnode">unified balance</Tooltip> within the Yellow Network. This is a purely off-chain operation executed as a database transaction on <Tooltip content="A virtual ledger layer that provides a unified ledger (through Nitro RPC) and coordinates state channels (through Nitrolite), providing chain abstraction for developers and users">the clearnode</Tooltip>, resulting in instant settlement (typically under 1 second) with zero gas fees. The transfer updates internal ledger entries using double-entry bookkeeping principles and creates a transaction record for both parties. Both sender and recipient must be connected to the same <Tooltip content="A virtual ledger layer that provides a unified ledger (through Nitro RPC) and coordinates state channels (through Nitrolite), providing chain abstraction for developers and users">clearnode</Tooltip> (have open channels). The security guarantee comes from the underlying on-chain channels that back the <Tooltip content="An abstraction that aggregates a user's funds across multiple blockchain networks, managed by Clearnode">unified balance</Tooltip>.

Reviewer: "(typically under 1 second) with zero gas fees" can be removed

Author: [1eeca228bfcc771893fae0befb9d0101c0bd888f](https://github.com/layer-3/docs/pull/69/commits/1eeca228bfcc771893fae0befb9d0101c0bd888f)

### When to Use

When sending funds to another Yellow Network user who is connected to the same <Tooltip content="A virtual ledger layer that provides a unified ledger (through Nitro RPC) and coordinates state channels (through Nitrolite), providing chain abstraction for developers and users">clearnode</Tooltip>. Common use cases include peer-to-peer payments, merchant payments, tipping, and rapid fund redistribution between users.

### Prerequisites

- Sender must be [authenticated](./authentication)
- Sender must have sufficient available balance (not locked in <Tooltip content="Off-chain channels built on top of payment channels, intended to be used by app developers to enable application-specific interactions and transactions without touching the blockchain">app sessions</Tooltip>)

> - Recipient must have at least one open <Tooltip content="A secure communication pathway between participants that locks funds in an on-chain smart contract while enabling off-chain state updates">channel</Tooltip> with <Tooltip content="A virtual ledger layer that provides a unified ledger (through Nitro RPC) and coordinates state channels (through Nitrolite), providing chain abstraction for developers and users">the clearnode</Tooltip> (be connected)

Reviewer 2: No, you can have a positive unified balance on Clearnode even if you don't have any open channel

Author: [1eeca228bfcc771893fae0befb9d0101c0bd888f](https://github.com/layer-3/docs/pull/69/commits/1eeca228bfcc771893fae0befb9d0101c0bd888f)

> - Recipient must have at least one open <Tooltip content="A secure communication pathway between participants that locks funds in an on-chain smart contract while enabling off-chain state updates">channel</Tooltip> with <Tooltip content="A virtual ledger layer that provides a unified ledger (through Nitro RPC) and coordinates state channels (through Nitrolite), providing chain abstraction for developers and users">the clearnode</Tooltip> (be connected)

Reviewer: "having an open channel" is not equal to ("be connected").  Anyways, Neither of these are required for the recipient.

Author: [1eeca228bfcc771893fae0befb9d0101c0bd888f](https://github.com/layer-3/docs/pull/69/commits/1eeca228bfcc771893fae0befb9d0101c0bd888f)

- Recipient must be identified by valid <Tooltip content="An entity (identified by a wallet address) that is part of a channel">wallet address</Tooltip> or user tag

### Request

| Parameter | Type | Required | Description | Format | Example | Notes |
|-----------|------|----------|-------------|--------|---------|-------|

> | `destination` | string (wallet address) | Yes (if destination_user_tag not provided) | Recipient's wallet address | 0x-prefixed hex string (20 bytes) | `"0x8B3192f2F7b1b34f2e4e7B8C9D1E0F2A3B4C5D6E"` | Recipient must be connected to this clearnode |

Reviewer 2: ```suggestion
| `destination` | string (wallet address) | Yes (if `destination_user_tag` not provided) | Recipient's wallet address | 0x-prefixed hex string (20 bytes) | `"0x8B3192f2F7b1b34f2e4e7B8C9D1E0F2A3B4C5D6E"` | Recipient must be connected to this clearnode |
```

Author: [1eeca228bfcc771893fae0befb9d0101c0bd888f](https://github.com/layer-3/docs/pull/69/commits/1eeca228bfcc771893fae0befb9d0101c0bd888f)

> | `destination_user_tag` | string | Yes (if destination not provided) | Recipient's human-readable identifier | Alphanumeric string | `"alice.yellow"` | Alternative to address; internal feature, may change |

Reviewer: Actually, destination_user_tag is not Recipient's human-readable identifier - its a randomly generated set of chars

Author: [1eeca228bfcc771893fae0befb9d0101c0bd888f](https://github.com/layer-3/docs/pull/69/commits/1eeca228bfcc771893fae0befb9d0101c0bd888f)

> | `allocations` | Allocation[] | Yes (minimum: 1) | Assets and amounts to transfer<br/><br/>**Structure (per allocation)**:<br/>• `asset` (string, required) - Asset identifier<br/>&nbsp;&nbsp;- Examples: "usdc", "eth", "weth", "btc"<br/>&nbsp;&nbsp;- Note: Must be lowercase; see get_assets for supported assets<br/>• `amount` (string, required) - Amount to transfer in human-readable format<br/>&nbsp;&nbsp;- Format: Decimal string (e.g., "50.0" for 50 USDC)<br/>&nbsp;&nbsp;- Note: Clearnode handles conversion to smallest unit internally | See structure | `[{"asset": "usdc", "amount": "50.0"}, {"asset": "eth", "amount": "0.01"}]` | Can send multiple assets in single transfer |

Reviewer 2: Would be good to have such structures described on the dedicated separate page and just reference them here

Author: Need to have a story for this

| `note` | string | No | Optional memo or note for the transfer | Max length: 256 characters | `"Payment for coffee"` | Visible to both parties in transaction history |

> ### Response

Reviewer 2: Review [this file](https://github.com/erc7824/nitrolite/blob/c708013f7bb2436f2ff7112be5e58f2d6058d79d/clearnode/pkg/rpc/api.go#L422-L424) and change the table accordingly

Author: [1eeca228bfcc771893fae0befb9d0101c0bd888f](https://github.com/layer-3/docs/pull/69/commits/1eeca228bfcc771893fae0befb9d0101c0bd888f)

> | Parameter | Type | Description | Format/Structure | Example | Notes |

Reviewer: This response is not related to the actual responce. Please refer to the [API doc](https://github.com/erc7824/nitrolite/blob/main/clearnode/docs/API.md#transfer-funds)
and describe the response structure acordingly

Author: [1eeca228bfcc771893fae0befb9d0101c0bd888f](https://github.com/layer-3/docs/pull/69/commits/1eeca228bfcc771893fae0befb9d0101c0bd888f)

|-----------|------|-------------|------------------|---------|-------|
| `transaction_id` | string | Unique identifier for this transfer | UUID or hash | `"txn_a1b2c3d4e5f6..."` | Use this to query transaction status |
| `status` | string | Transfer status | Value: "confirmed" (transfers are instant) | `"confirmed"` | - |
| `sender` | string (wallet address) | Sender's wallet address (authenticated user) | - | `"0x742d35Cc6634C0532925a3b844Bc9e7595f0bEb"` | - |
| `recipient` | string (wallet address) | Recipient's wallet address | - | `"0x8B3192f2F7b1b34f2e4e7B8C9D1E0F2A3B4C5D6E"` | Resolved from destination_user_tag if that was provided |
| `allocations` | Allocation[] | Confirmed transfer amounts | - | `[{"asset": "usdc", "amount": "50.0"}]` | - |

> | `sender_balance` | object | Sender's updated balance<br/><br/>**Structure**: asset: Balance per asset after transfer | See structure | `{"usdc": {"total": "50.0", "available": "50.0", "locked": "0"}}` | - |

Reviewer: We don't have neither of these fields

Author: [1eeca228bfcc771893fae0befb9d0101c0bd888f](https://github.com/layer-3/docs/pull/69/commits/1eeca228bfcc771893fae0befb9d0101c0bd888f)

| `timestamp` | number | Unix timestamp of transfer execution | - | `1699123456789` | - |
| `ledger_entries` | array | References to created ledger entries<br/><br/>**Structure**: [debitEntryId, creditEntryId] | See structure | - | For audit trail and reconciliation |

---

## Off-Chain Processing

When a transfer is executed, <Tooltip content="A virtual ledger layer that provides a unified ledger (through Nitro RPC) and coordinates state channels (through Nitrolite), providing chain abstraction for developers and users">the clearnode</Tooltip> performs the following operations:

```mermaid
sequenceDiagram
    box rgb(225,245,255) Client A
        participant Alice as Client A (Sender)
    end
    box rgb(255,225,245) Service
        participant Clearnode
    end
    box rgb(225,255,225) Client B
        participant Bob as Client B (Recipient)
    end

    Note over Alice: 1. Send Transfer Request
    Alice->>Clearnode: transfer(destination, allocations, note)

    Note over Clearnode: 2. Validate
    Clearnode->>Clearnode: Verify authentication
    Clearnode->>Clearnode: Check available balance

>     Clearnode->>Clearnode: Verify recipient exists

Reviewer 2: Right now you can transfer to a recipient that doesn't have anything on Clearnode

Author: [1eeca228bfcc771893fae0befb9d0101c0bd888f](https://github.com/layer-3/docs/pull/69/commits/1eeca228bfcc771893fae0befb9d0101c0bd888f)

>     Clearnode->>Clearnode: Verify recipient exists

Reviewer: We dont have "Verify recipient exists' step

Author: [1eeca228bfcc771893fae0befb9d0101c0bd888f](https://github.com/layer-3/docs/pull/69/commits/1eeca228bfcc771893fae0befb9d0101c0bd888f)

    Clearnode->>Clearnode: Validate allocations

    Note over Clearnode: 3. Update Ledger
    Clearnode->>Clearnode: Create debit entry (Alice -50 USDC)
    Clearnode->>Clearnode: Create credit entry (Bob +50 USDC)
    Clearnode->>Clearnode: Record transaction

    Note over Clearnode: 4. Send Responses

>     Clearnode->>Alice: {transaction_id, status: "confirmed", sender_balance}

Reviewer 2: ```suggestion
    Clearnode->>Alice: {ledger_transactions}
```

Author: [1eeca228bfcc771893fae0befb9d0101c0bd888f](https://github.com/layer-3/docs/pull/69/commits/1eeca228bfcc771893fae0befb9d0101c0bd888f)

>     Clearnode->>Bob: transfer_received notification

Reviewer: We dont have "transfer_received" notification. We have only transfer notification "tr". Please refer to the [doc](https://github.com/erc7824/nitrolite/blob/main/clearnode/docs/API.md#transfer-notifications)

Author: [1eeca228bfcc771893fae0befb9d0101c0bd888f](https://github.com/layer-3/docs/pull/69/commits/1eeca228bfcc771893fae0befb9d0101c0bd888f)

> 

Reviewer: Additionally, both parties receive Transfer and Balance Update notification

Author: [1eeca228bfcc771893fae0befb9d0101c0bd888f](https://github.com/layer-3/docs/pull/69/commits/1eeca228bfcc771893fae0befb9d0101c0bd888f)

    Note over Bob: 5. Balance Updated
    Bob->>Bob: Balance +50 USDC

```

### Step-by-Step Process

#### 1. Validates Request

<Tooltip content="A virtual ledger layer that provides a unified ledger (through Nitro RPC) and coordinates state channels (through Nitrolite), providing chain abstraction for developers and users">The clearnode</Tooltip> performs comprehensive validation:
- Verifies authentication and signature
- Checks sender has sufficient available balance

> - Verifies recipient exists (has open <Tooltip content="A secure communication pathway between participants that locks funds in an on-chain smart contract while enabling off-chain state updates">channel</Tooltip>)

Reviewer 2: remove

Author: [1eeca228bfcc771893fae0befb9d0101c0bd888f](https://github.com/layer-3/docs/pull/69/commits/1eeca228bfcc771893fae0befb9d0101c0bd888f)

> - Verifies recipient exists (has open <Tooltip content="A secure communication pathway between participants that locks funds in an on-chain smart contract while enabling off-chain state updates">channel</Tooltip>)

Reviewer: we dont verify that recipient exists

Author: [1eeca228bfcc771893fae0befb9d0101c0bd888f](https://github.com/layer-3/docs/pull/69/commits/1eeca228bfcc771893fae0befb9d0101c0bd888f)

- Validates allocations format and asset support

#### 2. Updates Ledger (Double-Entry Bookkeeping)

Every transfer creates two ledger entries:

**Debit Entry (Sender)**:
```
account: sender_address
asset: "usdc"
amount: -50.0
type: "DEBIT"

> related_account: recipient_address

Reviewer: We dont have type, "related_account: recipient_address
reference_id: transaction_id", amount is always positive and some other fields are different. Please refer to proper Go structure of a ledger entry, or dont even include it at all. 
[go ledger entry](https://github.com/erc7824/nitrolite/blob/c708013f7bb2436f2ff7112be5e58f2d6058d79d/clearnode/ledger.go#L18)

Author: [1eeca228bfcc771893fae0befb9d0101c0bd888f](https://github.com/layer-3/docs/pull/69/commits/1eeca228bfcc771893fae0befb9d0101c0bd888f)

reference_id: transaction_id
description: "Transfer to [recipient]"
```

**Credit Entry (Recipient)**:
```
account: recipient_address
asset: "usdc"
amount: +50.0
type: "CREDIT"
related_account: sender_address
reference_id: transaction_id
description: "Transfer from [sender]"
```

:::info Double-Entry Bookkeeping
The double-entry system ensures that the total of all debits always equals the total of all credits, providing mathematical proof of accounting accuracy. Every transaction is recorded twice - once as a debit and once as a credit.
:::

#### 3. Records Transaction

A user-facing transaction record is created:

```

> Ledger Transaction:

Reviewer: this struct is not related to actual structure.
Please use a correct structure or dont use it at all.
[LedgerTransaction](https://github.com/erc7824/nitrolite/blob/c708013f7bb2436f2ff7112be5e58f2d6058d79d/clearnode/ledger_transaction.go#L27)

Author: [1eeca228bfcc771893fae0befb9d0101c0bd888f](https://github.com/layer-3/docs/pull/69/commits/1eeca228bfcc771893fae0befb9d0101c0bd888f)

  transaction_id: unique_id
  type: "TRANSFER"
  sender: sender_address
  recipient: recipient_address
  asset: "usdc"
  amount: "50.0"
  status: "CONFIRMED"
  timestamp: current_time
```

#### 4. Sends Notifications

- **Sender** receives confirmation with updated balance

> - **Recipient** receives notification via WebSocket with balance update

Reviewer 2: ```suggestion
- **Recipient** receives notification on all of his connections with the balance update
```

Author: [1eeca228bfcc771893fae0befb9d0101c0bd888f](https://github.com/layer-3/docs/pull/69/commits/1eeca228bfcc771893fae0befb9d0101c0bd888f)

---

## Unified Balance Mechanics

The <Tooltip content="An abstraction that aggregates a user's funds across multiple blockchain networks, managed by Clearnode">unified balance</Tooltip> aggregates funds from all chains, providing a seamless cross-chain experience.

### Example: Multi-Chain Aggregation

```
User deposited:
  $10 USDC on Ethereum
  $5 USDC on Polygon
  $3 USDC on Base
  
Unified Balance: $18 USDC total

User can transfer: Any amount up to $18 USDC
Result: Transfer uses unified balance, no need to specify chain
```

> :::tip Cross-Chain Abstraction

Reviewer 2: Do not make too much attention on this concept, as it may actually change in the next version of the protocol

Author: [1eeca228bfcc771893fae0befb9d0101c0bd888f](https://github.com/layer-3/docs/pull/69/commits/1eeca228bfcc771893fae0befb9d0101c0bd888f)

Users don't need to know or care which chain their funds are on. Clearnode manages the underlying <Tooltip content="A secure communication pathway between participants that locks funds in an on-chain smart contract while enabling off-chain state updates">channel</Tooltip> allocations internally. When channels close, funds are distributed according to the final <Tooltip content="An abstraction that aggregates a user's funds across multiple blockchain networks, managed by Clearnode">unified balance</Tooltip> state.
:::

> ### Balance States

Reviewer: We dont have "Balance States". Funds are either in the unified account or not.
Better specify that there are 3 account types.
Main one is Unified Account, identified by wallet address
Another one is App Session Account identified by app session id, where participant wallets are beneficiaries of that account
The last one is channel escrow account, which is a temporaty account to lock funds if user requested some blockchain operation like resize, so that funds are temporary locked in this account until the transaction is confirmed on blockchain

Author: [1eeca228bfcc771893fae0befb9d0101c0bd888f](https://github.com/layer-3/docs/pull/69/commits/1eeca228bfcc771893fae0befb9d0101c0bd888f)

<Tooltip content="An abstraction that aggregates a user's funds across multiple blockchain networks, managed by Clearnode">Unified balance</Tooltip> has three components:

1. **Total**: Sum of available + locked

> 2. **Available**: Funds that can be transferred or withdrawn

Reviewer: remove all this

Author: [1eeca228bfcc771893fae0befb9d0101c0bd888f](https://github.com/layer-3/docs/pull/69/commits/1eeca228bfcc771893fae0befb9d0101c0bd888f)

3. **Locked**: Funds currently in <Tooltip content="Off-chain channels built on top of payment channels, intended to be used by app developers to enable application-specific interactions and transactions without touching the blockchain">app sessions</Tooltip>

**Only available balance can be transferred**.

---

> ## Security Considerations

Reviewer 2: @nksazonov @dimast-x what do you think about this section?)

> ## Security Considerations

Reviewer: I suggest removing Security Consideration section and write it properly once we are ready

Author: [1eeca228bfcc771893fae0befb9d0101c0bd888f](https://github.com/layer-3/docs/pull/69/commits/1eeca228bfcc771893fae0befb9d0101c0bd888f)

:::

### What If a Clearnode Disappears?

**Recovery Process**:
- Users can challenge-close their <Tooltip content="A secure communication pathway between participants that locks funds in an on-chain smart contract while enabling off-chain state updates">channel</Tooltip> with last known <Tooltip content="A snapshot of the channel at a point in time, including fund allocations and application-specific data">state</Tooltip>
- Funds are recovered according to last signed <Tooltip content="A snapshot of the channel at a point in time, including fund allocations and application-specific data">state</Tooltip>
- Worst case: lose value of un-checkpointed transfers

**Best Practice**:
- Periodically checkpoint important <Tooltip content="A snapshot of the channel at a point in time, including fund allocations and application-specific data">states</Tooltip> on-chain for large balances
- Keep local records of recent transfer history

---

## Error Cases

> | Error | Code | Description | Recovery |

Reviewer: we neither have these errors, nor err codes. Please use our actual errors or dont show it at all

Author: [1eeca228bfcc771893fae0befb9d0101c0bd888f](https://github.com/layer-3/docs/pull/69/commits/1eeca228bfcc771893fae0befb9d0101c0bd888f)

| **Authentication required** | 1001 | Not authenticated | Complete [authentication](./authentication) |
| **Insufficient available balance** | 1002 | Sender doesn't have enough | Check balance, wait for funds |
| **Recipient not found** | 1003 | No channels with clearnode | Verify recipient address, ensure they're connected |
| **Invalid destination address** | -32602 | Malformed wallet address | Verify address format (0x + 40 hex) |
| **Invalid allocation format** | -32602 | Negative amount or unsupported asset | Check allocation structure |
| **Amount exceeds available** | 1002 | Exceeds available (not total) balance | Reduce amount or close <Tooltip content="Off-chain channels built on top of payment channels, intended to be used by app developers to enable application-specific interactions and transactions without touching the blockchain">app sessions</Tooltip> |
| **Cannot transfer to self** | -32602 | sender == recipient | Use different recipient |
| **Asset not supported** | 1010 | Asset not available | Use `get_assets` to see supported assets |
| **Amount too small** | -32602 | Below minimum transfer amount | Increase amount |

---

> ## Transaction History

Reviewer 2: let's describe endpoint here the same way as all other, use only params definitions, as SDK is not a protocol, but implementation

Author: [1eeca228bfcc771893fae0befb9d0101c0bd888f](https://github.com/layer-3/docs/pull/69/commits/1eeca228bfcc771893fae0befb9d0101c0bd888f)

Users can query their transfer history using two methods:

### get_ledger_transactions

**Purpose**: User-facing transaction log with sender, recipient, amount, and type.

**Use Case**: Display transaction history in UI, export records.

```javascript
const transactions = await clearnode.call("get_ledger_transactions", {
  asset: "usdc",
  offset: 0,
  limit: 50
})

transactions.forEach(tx => {
  console.log(`${tx.tx_type}: ${tx.amount} ${tx.asset}`)
  console.log(`From: ${tx.from_account_tag || tx.from_account}`)
  console.log(`To: ${tx.to_account_tag || tx.to_account}`)
  console.log(`Note: ${tx.note || 'N/A'}`)
})
```

### get_ledger_entries

**Purpose**: Detailed accounting entries showing all debits and credits.

**Use Case**: Detailed reconciliation, audit trail, accounting integration.

```javascript
const entries = await clearnode.call("get_ledger_entries", {
  account_id: user_address,
  asset: "usdc"
})

entries.forEach(entry => {
  if (entry.debit > 0) {
    console.log(`Debit: -${entry.debit} ${entry.asset}`)
  } else {
    console.log(`Credit: +${entry.credit} ${entry.asset}`)
  }
})
```

---

## Implementation Notes

**Performance**:

> - Transfers are instant (< 1 second) and atomic

Reviewer: remove this

Author: [1eeca228bfcc771893fae0befb9d0101c0bd888f](https://github.com/layer-3/docs/pull/69/commits/1eeca228bfcc771893fae0befb9d0101c0bd888f)

- No blockchain transaction required
- Zero gas fees

**Network**:

> - Both parties must be online to receive notifications

Reviewer: this one can be removed

Author: [1eeca228bfcc771893fae0befb9d0101c0bd888f](https://github.com/layer-3/docs/pull/69/commits/1eeca228bfcc771893fae0befb9d0101c0bd888f)

- Transfer completes regardless of recipient online status
- Recipient sees updated balance when they reconnect

**Features**:
- <Tooltip content="An abstraction that aggregates a user's funds across multiple blockchain networks, managed by Clearnode">Unified balance</Tooltip> is updated immediately
- Transfer can include multiple assets in one operation

> - The transaction_id can be used to track and query the transfer status

Reviewer: we dont have a field "transaction_id".  Reference a correct structure.

Author: [1eeca228bfcc771893fae0befb9d0101c0bd888f](https://github.com/layer-3/docs/pull/69/commits/1eeca228bfcc771893fae0befb9d0101c0bd888f)

**Audit Trail**:
- Clearnode maintains complete audit trail of all transfers
- Double-entry bookkeeping ensures mathematical accuracy
- All records queryable via `get_ledger_*` methods

---

## Next Steps

Explore other off-chain operations:

- **[App Sessions](./app-sessions)** - Create multi-party application channels
- **[Queries & Notifications](./queries)** - Check balances, transactions, and receive updates
- **[Channel Methods](./channel-methods)** - Manage payment channels

For protocol fundamentals:

> - **[Authentication](./authentication)** - Understand session management

Reviewer: Understand authorization and session management

Author: done

Author: [2407a50f77b866e3cddc84621a441c69e6787515>](https://github.com/layer-3/docs/pull/69/commits/2407a50f77b866e3cddc84621a441c69e6787515)

import { tooltipDefinitions } from '@site/src/constants/tooltipDefinitions';

# Transfer Method

Transfer method enable instant, off-chain fund movement between users.

---

## Overview

The transfer system allows users to send funds to each other instantly using their <Tooltip content={tooltipDefinitions.unifiedBalance}>unified balance</Tooltip>, without any on-chain transactions. Transfers are backed by the security of underlying payment channels and use double-entry bookkeeping for accounting accuracy.

### Why Use Transfer?

**Instant Settlement**: Transfers complete immediately with instant finality.

**No Blockchain Fees**: No blockchain transactions means no gas costs for both sender and recipient.

**Cross-Chain Unified**: Send from your <Tooltip content={tooltipDefinitions.unifiedBalance}>unified balance</Tooltip> across multiple chains.

**Auditable**: Complete transaction history with double-entry ledger tracking.

:::success Instant Off-Chain Payments
Transfers provide the speed and convenience of traditional payment networks while maintaining the security guarantees of blockchain-backed channels.
:::

---

## transfer

### Name

`transfer`

### Usage

> Transfer funds from the authenticated user's <Tooltip content={tooltipDefinitions.unifiedBalance}>unified balance</Tooltip> to another user's <Tooltip content={tooltipDefinitions.unifiedBalance}>unified balance</Tooltip> within the Yellow Network. This is a purely off-chain operation executed as a database transaction on <Tooltip content={tooltipDefinitions.clearnode}>the clearnode</Tooltip>, resulting in instant settlement. The transfer updates internal ledger entries using double-entry bookkeeping principles and creates a transaction record for both parties. The security guarantee comes from the underlying on-chain channels that back the <Tooltip content={tooltipDefinitions.unifiedBalance}>unified balance</Tooltip>.

Reviewer: Replace "This is a purely off-chain operation executed as a database transaction on the clearnode, resulting in instant settlement.'
with "This is a purely off-chain operation executed as a database transaction on the clearnode, resulting in instant settlement."

Author: Changed it to: Transfer funds from the authenticated user's <Tooltip content={tooltipDefinitions.unifiedBalance}>unified balance</Tooltip> to another user's <Tooltip content={tooltipDefinitions.unifiedBalance}>unified balance</Tooltip> within the Yellow Network. This is a purely off-chain operation, which results in instant settlement. The transfer updates internal ledger entries using double-entry bookkeeping principles and creates a transaction record for both parties. The security guarantee comes from the underlying on-chain channels that back the <Tooltip content={tooltipDefinitions.unifiedBalance}>unified balance</Tooltip>.

Author: [aad0b80ca8e5c74d4c329a9ed21b7db470b0d336](https://github.com/layer-3/docs/pull/69/commits/aad0b80ca8e5c74d4c329a9ed21b7db470b0d336)

### When to Use

> When sending funds to another Yellow Network user. Common use cases include peer-to-peer payments, merchant payments, tipping, and rapid fund redistribution between users.

Reviewer: Remove "rapid fund redistribution between users", its a feature of app sessions, not transfers

Author: [aad0b80ca8e5c74d4c329a9ed21b7db470b0d336](https://github.com/layer-3/docs/pull/69/commits/aad0b80ca8e5c74d4c329a9ed21b7db470b0d336)

### Prerequisites

- Sender must be [authenticated](./authentication)
- Sender must have sufficient available balance in <Tooltip content={tooltipDefinitions.unifiedBalance}>unified account</Tooltip>
- Recipient must be identified by valid <Tooltip content={tooltipDefinitions.participant}>wallet address</Tooltip> or user tag

:::info Recipient Requirements

> The recipient does not need to have an existing balance or account on the clearnode. Transfers can be sent to any valid wallet address, and the recipient's account will be created automatically if it doesn't exist.

Reviewer: created automatically on the first login.

Author: [aad0b80ca8e5c74d4c329a9ed21b7db470b0d336](https://github.com/layer-3/docs/pull/69/commits/aad0b80ca8e5c74d4c329a9ed21b7db470b0d336)

:::

### Request

| Parameter | Type | Required | Description | Format | Example | Notes |
|-----------|------|----------|-------------|--------|---------|-------|
| `destination` | string (wallet address) | Yes (if `destination_user_tag` not provided) | Recipient's wallet address | 0x-prefixed hex string (20 bytes) | `"0x8B3192f2F7b1b34f2e4e7B8C9D1E0F2A3B4C5D6E"` | - |

> | `destination_user_tag` | string | Yes (if destination not provided) | Recipient's randomly generated user identifier | Alphanumeric string | `"UX123D"` | Alternative to address; internal feature, may change |

Reviewer: user tag is 6 chars long, not 5

Author: Already 6 ;D

When a transfer is executed, <Tooltip content={tooltipDefinitions.clearnode}>the clearnode</Tooltip> performs the following operations:

```mermaid
sequenceDiagram
    box rgb(225,245,255) Client A
        participant Alice as Client A (Sender)
    end
    box rgb(255,225,245) Service
        participant Clearnode
    end
    box rgb(225,255,225) Client B
        participant Bob as Client B (Recipient)
    end

    Note over Alice: 1. Send Transfer Request
    Alice->>Clearnode: transfer({ destination, allocations })

    Note over Clearnode: 2. Validate
    Clearnode->>Clearnode: Verify authentication
    Clearnode->>Clearnode: Check available balance
    Clearnode->>Clearnode: Validate allocations

    Note over Clearnode: 3. Update Ledger
    Clearnode->>Clearnode: Create debit entry (Alice -50 USDC)
    Clearnode->>Clearnode: Create credit entry (Bob +50 USDC)
    Clearnode->>Clearnode: Record transaction

    Note over Clearnode: 4. Send Responses & Notifications
    Clearnode->>Alice: {ledger_transactions}
    Clearnode->>Alice: bu (balance update) notification
    Clearnode->>Bob: tr (transfer) notification
    Clearnode->>Bob: bu (balance update) notification

>     Note over Bob: 5. Balance Updated

Reviewer: Client A also receives tr (transfer) notification.
also move "Clearnode->>Alice: {ledger_transactions}" to be the last step, and instead of ledger_transactions, simply say that this is "response"

Author: [aad0b80ca8e5c74d4c329a9ed21b7db470b0d336](https://github.com/layer-3/docs/pull/69/commits/aad0b80ca8e5c74d4c329a9ed21b7db470b0d336)

#### 1. Validates Request

<Tooltip content={tooltipDefinitions.clearnode}>The clearnode</Tooltip> performs comprehensive validation:
- Verifies authentication and signature
- Checks sender has sufficient available balance in <Tooltip content={tooltipDefinitions.unifiedBalance}>unified account</Tooltip>
- Validates allocations format and asset support

#### 2. Updates Ledger (Double-Entry Bookkeeping)

Every transfer creates two ledger entries - one for the sender and one for the recipient. The ledger uses double-entry bookkeeping principles where each entry has both `credit` and `debit` fields, with amounts always recorded as positive values.

:::info Double-Entry Bookkeeping
The double-entry system ensures that the total of all debits always equals the total of all credits, providing mathematical proof of accounting accuracy. Every transfer is recorded twice - once as a debit to the sender's account and once as a credit to the recipient's account.
:::

#### 3. Records Transaction

A user-facing transaction record is created for each asset being transferred, containing information about the sender, recipient, asset, and amount.

#### 4. Sends Notifications

> - **Sender** receives response with transaction details

Reviewer: move "**Sender** receives response with transaction details" into section 5. Response

Author: [aad0b80ca8e5c74d4c329a9ed21b7db470b0d336](https://github.com/layer-3/docs/pull/69/commits/aad0b80ca8e5c74d4c329a9ed21b7db470b0d336)

---

## Unified Balance Mechanics

The <Tooltip content={tooltipDefinitions.unifiedBalance}>unified balance</Tooltip> aggregates funds from all chains.

### Example: Multi-Chain Aggregation

```
User deposited:
  $10 USDC on Ethereum
  $5 USDC on Polygon
  $3 USDC on Base
  
Unified Balance: $18 USDC total

User can transfer: Any amount up to $18 USDC
```

### Account Types

The ledger system maintains three types of accounts:

1. **Unified Account**: Main account identified by wallet address. This is where user funds are stored and can be transferred or withdrawn.

2. **App Session Account**: Identified by app session ID. Participant wallets are beneficiaries of this account. Funds in app sessions are locked for the duration of the session.

3. **Channel Escrow Account**: Temporary account that locks funds when user requests blockchain operations like resize. Funds remain in this account until the transaction is confirmed on-chain.

---

## Error Cases

:::note Error Format
The protocol does not use numeric error codes. Errors are returned as method `"error"` with descriptive messages.
:::

Common error scenarios:

> | Error | Description | Recovery |

Reviewer: We don't have these errors. The errors must follow the actual errors in the code. You can remove this section for now, and leave a TODO (make sure the todo comment is not shown to users)

Author: [aad0b80ca8e5c74d4c329a9ed21b7db470b0d336](https://github.com/layer-3/docs/pull/69/commits/aad0b80ca8e5c74d4c329a9ed21b7db470b0d336)
